index.js

```diff
@@ -0,0 +1,377 @@
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--word', '--color', 'foo']))
+//. Right (Pair ({color: false, email: Nothing, words: ['--color']}) (['foo']))
+//. ```
+//.
+//. Option handlers may perform validation:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--email', 'xxx']))
+//. Left ('"xxx" is not a valid email address')
+//. ```
+//.
+//. Specifying an option multiple times is supported, and each handler defines
+//. its own semantics:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['-c', '-c', '-c', '-c', '-c']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-e', 'a@x.org', '-e', 'b@x.org']))
+//. Right (Pair ({color: false, email: Just ('b@x.org'), words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-w', 'foo', '-w', 'bar']))
+//. Right (Pair ({color: false, email: Nothing, words: ['foo', 'bar']}) ([]))
+//. ```
+//.
+//. As soon as an argument is encountered that is not a flag or an option
+//. value, that argument and any subsequent arguments are returned as the
+//. positional arguments (along with the final configuration value):
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', 'foo', 'bar']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['foo', 'bar']))
+//. ```
+//.
+//. `--` can be used to indicate that the remaining arguments are positional:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', '--', '--no-color']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['--no-color']))
+//. ```
+//.
+//. An unrecognized flag/option name results in an error message:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['-x', '-y', '-z']))
+//. Left ('-x is unrecognized')
+//.
+//. > parseArgs (spec) (Pair (conf) (['--xxx', '--yyy', '--zzz']))
+//. Left ('--xxx is unrecognized')
+//. ```
+//.
+//. Here is the type of `parseArgs` with type aliases expanded:
+//.
+//. ```haskell
+//. parseArgs :: StrMap (Either (a -> a) (String -> Either String (a -> a)))
+//.           -> Pair a (Array String)
+//.           -> Either String (Pair a (Array String))
+//. ```
+//.
+//. Although names can improve readability, the following example demonstrates
+//. that this module's functionality is exposed via a single, pure function:
+//.
+//. ```javascript
+//. > parseArgs
+//. .   ({'--foo': Left (conf => ({...conf, foo: true})),
+//. .     '--bar': Right (bar => Right (conf => ({...conf, bar: Just (bar)})))})
+//. .   (Pair ({foo: false, bar: Nothing})
+//. .         (['--foo', '--bar', 'baz', 'quux']))
+//. Right (Pair ({foo: true, bar: Just ('baz')}) (['quux']))
+//. ```
+const parseArgs =
+def ('parseArgs')
+    ({})
+    ([$.StrMap (Handler (a)),
+      $.Pair (a) ($.Array ($.String)),
+      $.Either ($.String) ($.Pair (a) ($.Array ($.String)))])
+    (spec => pair => {
+       let conf = pair.fst;
```

Comment:
It looks like the indentation here is off by one space. 

![latest-screenshot](https://user-images.githubusercontent.com/1217745/132818306-22364dd9-3528-46ce-abc3-b093d0eff97a.png)

:point_up: 
2 - 2 - 3

Reply:
Observant! :eyes:

This indentation is intentional. Consider the following code:

```javascript
Future.fork (s => {
               process.stderr.write (s + '\n');
               process.exit (1);
             })
            (s => {
               process.stdout.write (s);
               process.exit (0);
             })
            (program);
```

Aligning `}` with `(` would feel wrong to me just as the following code does:

```javascript
S.unwords (['foo',
           'bar',
           'baz'])
```

I view everything between `[` and `]` as a region. Base indentation within this region should be consistent:

```
S.unwords ([xxxxxx
            xxxxxx
            xxxxx])
✅
```

```
S.unwords ([xxxxxx
           xxxxxx
           xxxxx])
❌
```

I view an arrow function *in its entirety* as a region. Base indentation within this region should be consistent (with the body indented by an additional two spaces):

```
Future.fork (xxxxxx
             xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
             xxxxxxxxxxxxxxxxxxx
             x)
            (oooooo
             ooooooooooooooooooooooooooo
             ooooooooooooooooooo
             o)
            (program);
✅
```

```
Future.fork (xxxxxx
            xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
            xxxxxxxxxxxxxxxxxxx
            x)
            (oooooo
            ooooooooooooooooooooooooooo
            ooooooooooooooooooo
            o)
            (program);
❌
```

You may not share my view, but hopefully it seems internally consistent. You will see this style used in other Sanctuary projects too. :)

Reply:
Okay, no problem then. :)